cloudinit/cmd/main.py

```diff
@@ -280,7 +280,7 @@ def main_init(name, args):
     mode = sources.DSMODE_LOCAL if args.local else sources.DSMODE_NETWORK
 
     if mode == sources.DSMODE_NETWORK:
-        existing = "trust"
+        existing = "check"
```

Comment:
`mode` is only `local` if `args.local` is set. In your case, you use Config Drive as data source, and it is located on `/dev/sr0`, so why you need to modify `existing = "check"` under `if mode == sources.DSMODE_NETWORK:`, is that correct? I don't fully understand this part of the code, please forgive my dumb question here.

Also, you're setting it to `check`, meaning cloud-init will never trust a DS if `mode` is `sources.DSMODE_NETWORK`. IIUC, this adds an extra check to the function `_restore_from_checked_cache()` that will check if `check_instance_id` key is present and will `ds.check_instance_id(self.cfg)` as well before restoring DS from cache. Is that ok considering that that's gonna happen for all DS which `mode` is `sources.DSMODE_NETWORK`?

Reply:
It will be helpful to read all of the discussion on the original PR:

https://github.com/canonical/cloud-init/pull/229

I have to keep re-reading it to understand the expect behavior change.

In particular, I was not able to successfully confirm that these changes provide the described behavior:  On subsequent boots, if the datasource is not found, cloud-init can attempt to re-use the existing datasource (found at /var/lib/cloud/instance/$previous_instance_id).

On Ec2, I was never able to observe it doing this.  

So before we land this, I'd like to see:

1) a unittest which walks cloud-init through this path verifying that if none of the configured datasources are found *and* there's a previous datasource (and the object we restore matches the previous-instance id value) then we'll use that.  This set of unittests should cover both local only (like ConfigDrive) and local/net datasources.  

One previous concern I had was that on Openstack, we could end up crawling IMDS twice;  I think that's not the case; but we should verify that.

2) I'd also want confirmation that with this patch applied it does improve behavior.  We have 3 potential scenarios:

  a) Ec2 where IMDS is not available (I've done initial testing, but was not able to confirm this works there)
  b) OVF where the ISO is removed on subsequent boots
  c) ConfigDrive where ISO is removed on subsequent boots

And I think success here (for at least b and c) was that cloud-init resumes the previous instance-id which preserves hots ssh keys and existing network config.

Reply:
> `mode` is only `local` if `args.local` is set. In your case, you use Config Drive as data source, and it is located on `/dev/sr0`, so why you need to modify `existing = "check"` under `if mode == sources.DSMODE_NETWORK:`, is that correct? I don't fully understand this part of the code, please forgive my dumb question here.
> 
> Also, you're setting it to `check`, meaning cloud-init will never trust a DS if `mode` is `sources.DSMODE_NETWORK`. IIUC, this adds an extra check to the function `_restore_from_checked_cache()` that will check if `check_instance_id` key is present and will `ds.check_instance_id(self.cfg)` as well before restoring DS from cache. Is that ok considering that that's gonna happen for all DS which `mode` is `sources.DSMODE_NETWORK`?

This PR I have reopend from #229. Even I had to go through the initial change in #229 many times to understand it.  As @raharper explained #229 has initial discussion.
With this PR I am able to confirm its not resetting to fallback dhcp network when config drive is removed on PowerVM hypervisors setup.  To answer your question though, as per the documentation in the below link [1] here with "trust" changing to "check" I believe it forces to determine that its first boot or subsequent boot instead of trusting unconditionally. I hope my understanding is right here @raharper 

existing = "check": 
cloud-init attempts to determine which case it is running in by checking the instance ID in the cache against the instance ID it determines at runtime. If they do not match, then this is an instance’s first boot; otherwise, it’s a subsequent boot.

existing = "trust" : 
trust the instance ID that is present in the system unconditionally. This means that cloud-init will never detect a new instance when the cache is present, and it follows that the only way to cause cloud-init to detect a new instance (and therefore its first boot) is to manually remove cloud-init’s cache. 

[1] https://cloudinit.readthedocs.io/en/latest/topics/boot.html

Reply:
> a) Ec2 where IMDS is not available (I've done initial testing, but was not able to confirm this works there)
> b) OVF where the ISO is removed on subsequent boots
> c) ConfigDrive where ISO is removed on subsequent boots

@raharper  sure will work on unit tests suggested. if I have read your comment right, with this PR, you still think "a)" is not working but "b)" and "c)" seems to be successful. Am I right in my understanding?

I am very sure about the issue with Ec2 where IMDS is not available. I need to understand more on this.  could you please help me to test a) I mean currently any specific UT you are running to test a)? so with this PR I am kind trying to address the configdrive issue for subsequent boot.

Reply:
> With this PR I am able to confirm its not resetting to fallback dhcp network when config drive is removed on PowerVM hypervisors setup.

Great!  I'd love to see the /var/log/cloud-init.log from that test.

> To answer your question though, as per the documentation in the below link [1] here with "trust" changing to "check" I believe it forces to determine that its first boot or subsequent boot instead of trusting unconditionally. I hope my understanding is right here @raharper

Yes.   This check answers the question: does cloud-init second stage  (1. cloud-init init --local, 2. cloud-init (net) 3. cloud-config modules 4. cloud-config --final) and beyond "trust" that the instance_id on the obj.pkl (/var/lib/cloud/instance/obj.pkl) is the same as the *actual* instance-id (datasource.get_instance_id())?   This trust value is needed due to the fact that cloud-init is exec'ed 4 times; so stages 2 through 4 need to restore state; this is done via the obj.pkl being loaded (pickle.load()) and avoids running code we've already run  more than once.

For local datasources (like ConfigDrive or NoCloud); checking both at stage 1 and 2 has little impact.  My original concern was for datasources which read metadata from the network.  In particular Ec2 datasource runs at local time reading from the network to obtain instance metadata.  Adding a second ds check for stage two, I thought, would mean we'd crawl IMDS a second time.  However, in practice, I did not observe that since cloud-init always writes out an obj.pkl between stages.  I expected Ec2 to hit IMDS a second time; however, I think what happens is that if local writes out obj.pkl then when stage 2 runs, we see this code flow:

```
    def _restore_from_checked_cache(self, existing):
        if existing not in ("check", "trust"):
            raise ValueError("Unexpected value for existing: %s" % existing)

        ds = self._restore_from_cache()   # <---- we load the obj.pkl here
        if not ds:
            return (None, "no cache found")

        run_iid_fn = self.paths.get_runpath('instance_id') <-- this symlink points to current instance-id which stage 1 (local sets up)
        if os.path.exists(run_iid_fn):
            run_iid = util.load_file(run_iid_fn).strip()
        else:
            run_iid = None

        if run_iid == ds.get_instance_id():  # <--- this method on Ec2 has attributes set with the instance-id so we return a valid DS here
            return (ds, "restored from cache with run check: %s" % ds)
        elif existing == "trust":
            return (ds, "restored from cache: %s" % ds)
        else:
            if (hasattr(ds, 'check_instance_id') and
                    ds.check_instance_id(self.cfg)):
                return (ds, "restored from checked cache: %s" % ds)
            else:
                return (None, "cache invalid in datasource: %s" % ds)

```

The second case I was worried about was OpenStack, however, the run-iid and
get_instance_id() also succeeds in returning a DS from the restored obj.pkl
created at stage 1.

```
    def _restore_from_checked_cache(self, existing):
        if existing not in ("check", "trust"):
            raise ValueError("Unexpected value for existing: %s" % existing)

        ds = self._restore_from_cache()   # <---- we load the obj.pkl here
        if not ds:
            return (None, "no cache found")

        run_iid_fn = self.paths.get_runpath('instance_id') <-- this symlink points to current instance-id which stage 1 (local sets up)
        if os.path.exists(run_iid_fn):
            run_iid = util.load_file(run_iid_fn).strip()
        else:
            run_iid = None

        if run_iid == ds.get_instance_id():  # <--- this method on OpenStack returns ds.metadata['instance-id'] and we return valid DS here
            return (ds, "restored from cache with run check: %s" % ds)
        elif existing == "trust":
            return (ds, "restored from cache: %s" % ds)
        else:
            if (hasattr(ds, 'check_instance_id') and
                    ds.check_instance_id(self.cfg)):
                return (ds, "restored from checked cache: %s" % ds)
            else:
                return (None, "cache invalid in datasource: %s" % ds)

```

Now, if at local time (stage 1) we cannot reach our datasource (on the network
for openstack/ec2) then no new obj.pkl is created (but one should exist from
the previous boot) and stage 2 should raise DataSourceNotFound, then load the
obj.pkl and compare the restored instance-id to the symlink of previous-iid.

This is the path I'd like to capture in unittests.

Reply:
> > a) Ec2 where IMDS is not available (I've done initial testing, but was not able to confirm this works there)
> > b) OVF where the ISO is removed on subsequent boots
> > c) ConfigDrive where ISO is removed on subsequent boots
>
> @raharper sure will work on unit tests suggested. if I have read your
> comment right, with this PR, you still think "a)" is not working but "b)"
> and "c)" seems to be successful. Am I right in my understanding?

Yes; that was my concern.   Here:

https://github.com/canonical/cloud-init/pull/229#discussion_r397500586

There was sort of an oddity.  If IMDS is down when cloud-init --local runs
*and* there isn;t an obj.pkl (which implies a first boot); then this PR
doesn't help; there is nothing to restore.

Then when there *is* an obj.plk in /var/lib/cloud/instances/<previous iid>
and /var/lib/cloud/instance (symlink) still points to this dir ^

I found that cloud-init ended up using the obj.pkl anyway without triggering
the DataSourceNotFound path.

I want to understand why that;s the case.  It could be due to how I was
testing;  I think I may have not actually simulated a reboot of the same
instance with no IMDS properly.

>
> I am very sure about the issue with Ec2 where IMDS is not available. I need
> to understand more on this. could you please help me to test a) I mean
> currently any specific UT you are running to test a)? so with this PR I am
> kind trying to address the configdrive issue for subsequent boot.

For a) I only did runtime experimenting.  I checked out this PR and built it
as a package, and upgraded cloud-init on an instance.  My test was:

sudo iptables -A OUTPUT -d 169.254.169.254 -j DROP  # simulate disabled IMDS
cloud-init --debug init --local
cloud-init --debug init

and I would look at cloud-init.log to see if we took the DatasourceNotFound
execption.

However, I believe we need to remove the instance-id symlink in
/run/cloud-init/.instance-id;  I didn't do that and I think that prevented
from simulating the path we'd expect on a reboot where IMDS was not working.


As for unittests of this;  I think adding something to

tests/unittests/test_runs/test_simple.py

It will takes some effort but we want to run the same code paths that
cloud-init init --local and cloud-init init do; then we'll simulate a reboot
by removing any of the cloud-init content from the rundir path.

Reply:
> if run_iid == ds.get_instance_id():  # <--- this method on Ec2 has attributes set with the instance-id so we return a valid DS here
>  if run_iid == ds.get_instance_id():  # <--- this method on OpenStack returns ds.metadata['instance-id']

I was wrong about this;  and this is why I didn't see Ec2 hit the DatasourceNotFound path; I forgot to clean /run/cloud-init/.instance-id and this short-circuited the code path.

Reply:
> > With this PR I am able to confirm its not resetting to fallback dhcp network when config drive is removed on PowerVM hypervisors setup.
> 
> Great! I'd love to see the /var/log/cloud-init.log from that test.
> 
> > To answer your question though, as per the documentation in the below link [1] here with "trust" changing to "check" I believe it forces to determine that its first boot or subsequent boot instead of trusting unconditionally. I hope my understanding is right here @raharper
> 
> Yes. This check answers the question: does cloud-init second stage (1. cloud-init init --local, 2. cloud-init (net) 3. cloud-config modules 4. cloud-config --final) and beyond "trust" that the instance_id on the obj.pkl (/var/lib/cloud/instance/obj.pkl) is the same as the _actual_ instance-id (datasource.get_instance_id())? This trust value is needed due to the fact that cloud-init is exec'ed 4 times; so stages 2 through 4 need to restore state; this is done via the obj.pkl being loaded (pickle.load()) and avoids running code we've already run more than once.
> 
> For local datasources (like ConfigDrive or NoCloud); checking both at stage 1 and 2 has little impact. My original concern was for datasources which read metadata from the network. In particular Ec2 datasource runs at local time reading from the network to obtain instance metadata. Adding a second ds check for stage two, I thought, would mean we'd crawl IMDS a second time. However, in practice, I did not observe that since cloud-init always writes out an obj.pkl between stages. I expected Ec2 to hit IMDS a second time; however, I think what happens is that if local writes out obj.pkl then when stage 2 runs, we see this code flow:
> 
> ```
>     def _restore_from_checked_cache(self, existing):
>         if existing not in ("check", "trust"):
>             raise ValueError("Unexpected value for existing: %s" % existing)
> 
>         ds = self._restore_from_cache()   # <---- we load the obj.pkl here
>         if not ds:
>             return (None, "no cache found")
> 
>         run_iid_fn = self.paths.get_runpath('instance_id') <-- this symlink points to current instance-id which stage 1 (local sets up)
>         if os.path.exists(run_iid_fn):
>             run_iid = util.load_file(run_iid_fn).strip()
>         else:
>             run_iid = None
> 
>         if run_iid == ds.get_instance_id():  # <--- this method on Ec2 has attributes set with the instance-id so we return a valid DS here
>             return (ds, "restored from cache with run check: %s" % ds)
>         elif existing == "trust":
>             return (ds, "restored from cache: %s" % ds)
>         else:
>             if (hasattr(ds, 'check_instance_id') and
>                     ds.check_instance_id(self.cfg)):
>                 return (ds, "restored from checked cache: %s" % ds)
>             else:
>                 return (None, "cache invalid in datasource: %s" % ds)
> ```
> 
> The second case I was worried about was OpenStack, however, the run-iid and
> get_instance_id() also succeeds in returning a DS from the restored obj.pkl
> created at stage 1.
> 
> ```
>     def _restore_from_checked_cache(self, existing):
>         if existing not in ("check", "trust"):
>             raise ValueError("Unexpected value for existing: %s" % existing)
> 
>         ds = self._restore_from_cache()   # <---- we load the obj.pkl here
>         if not ds:
>             return (None, "no cache found")
> 
>         run_iid_fn = self.paths.get_runpath('instance_id') <-- this symlink points to current instance-id which stage 1 (local sets up)
>         if os.path.exists(run_iid_fn):
>             run_iid = util.load_file(run_iid_fn).strip()
>         else:
>             run_iid = None
> 
>         if run_iid == ds.get_instance_id():  # <--- this method on OpenStack returns ds.metadata['instance-id'] and we return valid DS here
>             return (ds, "restored from cache with run check: %s" % ds)
>         elif existing == "trust":
>             return (ds, "restored from cache: %s" % ds)
>         else:
>             if (hasattr(ds, 'check_instance_id') and
>                     ds.check_instance_id(self.cfg)):
>                 return (ds, "restored from checked cache: %s" % ds)
>             else:
>                 return (None, "cache invalid in datasource: %s" % ds)
> ```
> 
> Now, if at local time (stage 1) we cannot reach our datasource (on the network
> for openstack/ec2) then no new obj.pkl is created (but one should exist from
> the previous boot) and stage 2 should raise DataSourceNotFound, then load the
> obj.pkl and compare the restored instance-id to the symlink of previous-iid.
> 
> This is the path I'd like to capture in unittests.

Attached the logs in the LP: 1893770